File: host/completion/scoped-reference.mjs
import { ScopedCompletionGuard } from './scoped-guard.mjs';
const [mode,identity,oldWindow,oldOperation]=process.argv.slice(2);
const g=new ScopedCompletionGuard(identity),window=g.acquire(1),operation=g.submit(window);
if(mode==='create') console.log(JSON.stringify({window,operation}));
else {
  const foreign=[];
  for(const attempt of [()=>g.submit(oldWindow),()=>g.read(oldWindow),()=>g.release(oldWindow),()=>g.poll(oldOperation),()=>g.cancel(oldOperation),()=>g.complete(oldOperation,[9])]) {
    try {attempt();foreign.push(0);}catch(error){foreign.push(error);}
  }
  const live=g.counts();g.complete(operation,[7]);const bytes=g.read(window);g.release(operation);g.release(window);
  console.log(JSON.stringify({window,operation,foreign,live,bytes,final:g.counts()}));
}

File: host/completion/rust/src/bin/scoped-reference.rs
use serde_json::json;
use wasmc_completion_guard::scoped::{BindingIdentity, ScopedCompletionGuard, ScopedToken};
fn main() {
    let args: Vec<_> = std::env::args().collect();
    let mut g = ScopedCompletionGuard::new(BindingIdentity::from_hex(&args[2]).unwrap()).unwrap();
    let w = g.acquire(1).unwrap();
    let op = g.submit(&w).unwrap();
    if args[1] == "create" {
        println!("{}", json!({"window":w.to_wire(),"operation":op.to_wire()}));
        return;
    }
    let oldw = ScopedToken::from_wire(&args[3]).unwrap();
    let oldop = ScopedToken::from_wire(&args[4]).unwrap();
    let foreign = [
        g.submit(&oldw).err(),
        g.read(&oldw).err(),
        g.release(&oldw).err(),
        g.poll(&oldop).err(),
        g.cancel(&oldop).err(),
        g.complete(&oldop, &[9], 0).err(),
    ];
    let live = g.counts();
    g.complete(&op, &[7], 0).unwrap();
    let bytes = g.read(&w).unwrap();
    g.release(&op).unwrap();
    g.release(&w).unwrap();
    println!(
        "{}",
        json!({"window":w.to_wire(),"operation":op.to_wire(),"foreign":foreign,"live":live,"bytes":bytes,"final":g.counts()})
    );
}
